volume/image.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/hashicorp/go-multierror"
+	"github.com/opencontainers/image-spec/identity"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type imageVolumeProvider struct {
+	client      *containerd.Client
+	image       string
+	target      string
+	snapshotter string
+	snapshot    string
+}
+
+// ImageOpt allows setting optional properties of the provider.
+type ImageOpt func(*imageVolumeProvider)
+
+// WithSnapshotter sets the snapshotter to pull images.
+func WithSnapshotter(ss string) ImageOpt {
+	return func(p *imageVolumeProvider) {
+		p.snapshotter = ss
+	}
+}
+
+func getV1ImageConfig(ctx context.Context, image containerd.Image) (*v1.Image, error) {
+	var result v1.Image
+
+	ic, err := image.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	switch ic.MediaType {
+	case v1.MediaTypeImageConfig, images.MediaTypeDockerSchema2Config:
+		p, err := content.ReadBlob(ctx, image.ContentStore(), ic)
+		if err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(p, &result); err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unknown type: %s", ic.MediaType)
+	}
+	return &result, nil
+}
+
+// FromImage returns a new provider to that expose the volumes on the given image.
+func FromImage(client *containerd.Client, image, snapshot string, opts ...ImageOpt) Provider {
+	p := imageVolumeProvider{
+		snapshotter: containerd.DefaultSnapshotter,
+		client:      client,
+		image:       image,
+		snapshot:    snapshot,
+	}
+
+	for _, opt := range opts {
+		opt(&p)
+	}
+
+	return &p
+}
+
+func (p *imageVolumeProvider) Name() string {
+	return p.image
+}
+
+func (p *imageVolumeProvider) Delete(ctx context.Context) error {
+	var retErr error
+
+	if p.target != "" {
+		retErr = mount.UnmountAll(p.target, 0)
+	}
+
+	ss := p.client.SnapshotService(p.snapshotter)
+	err := ss.Remove(ctx, p.snapshot)
+	if err != nil {
+		retErr = multierror.Append(retErr, err)
+	}
+	defer ss.Close()
+
+	return retErr
+}
+
+func (p *imageVolumeProvider) CreateVolumesUnder(ctx context.Context, tempDir string) ([]*Volume, error) {
+	image, err := p.client.GetImage(ctx, p.image)
+	if err != nil {
+		return nil, err
+	}
+
+	unpacked, err := image.IsUnpacked(ctx, p.snapshotter)
+	if err != nil {
+		return nil, err
+	}
+
+	if !unpacked {
+		err := image.Unpack(ctx, p.snapshotter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	config, err := getV1ImageConfig(ctx, image)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(config.Config.Volumes) == 0 {
+		return nil, nil
+	}
+
+	root, err := p.mount(ctx, tempDir)
+	if err != nil {
+		return nil, err
+	}
+	p.target = root
+
+	result := make([]*Volume, 0, len(config.Config.Volumes))
+	for path := range config.Config.Volumes {
+		result = append(result, &Volume{hostPath: filepath.Join(root, path), containerPath: path})
```

Comment:
I'm pretty sure this `filepath.Join` is [CVE-2022-23648](https://github.com/containerd/containerd/security/advisories/GHSA-crp2-qrr5-8pq7) :)

We should make the same change as the CRI plugin and probably also add a test to make sure that image volumes can't read from the host.

Reply:
Right. The fix was https://github.com/containerd/containerd/commit/8acbb276477c8353ddb7551b58cb3fdeb0b13caa. Let me check where the fs is coming.

Reply:
Fixed in the last revision.

Reply:
Don't we need `fs.RootPath` here as well?